### DISCUSSION TITLE
difference() has different behaviours with tube() than brick() meshes

### DISCUSSION THREAD

**Discussion by jlvila** [jimy-byerley/pymadcad#27] [Q&A] [answered]
Dear,
I had a simple function to carry out a difference() between 2 meshes.
### Case 1
`difference(brickA, brickB)` (red minus blue)
![rect1](https://user-images.githubusercontent.com/8035931/157838252-9172bc6c-d02b-4e3c-9a1b-5b7190c62e84.png)
The result is the expected one
![Difference](https://user-images.githubusercontent.com/8035931/157838493-4e1268d7-315a-4482-983c-40526170197d.png)
### Case 2
`difference(brickA, tubeB)` (red minus "tube")
![tube1](https://user-images.githubusercontent.com/8035931/157838663-feb20b7d-2b8b-4bf8-a4c2-8eeacbc5316a.png)
But here the result is not the expected one!
![Difference_tube](https://user-images.githubusercontent.com/8035931/157838760-cb394d1b-aba5-4b29-aea3-ece197071327.png)

### The code
   Context: w10, python 3.9.7, pyMadCad v0.11

   The shapes are simply built with `tube()` and `brick()` methods.

   ```python
   def difference(self, aKey, bKey):
        self.shapes3d[aKey].mergeclose()
        self.shapes3d[bKey].mergeclose()
        _diff = difference(
            self.shapes3d[aKey],
            self.shapes3d[bKey])
        _color = self.shapes3d[aKey].options["color"]
        _diff.option(color=_color)
        _diff.stripgroups()
        _diff.strippoints()
        _diff.mergeclose()
        _diff.finish()
        self.scene3d.scene.add(_diff, key=aKey)
        self.shapes3d[aKey] = _diff
        npart4 = note_bounds(_diff)
        self.scene3d.scene.add(npart4, f'{key2}_bounds')
        self.scene3d.scene.dequeue()
        self.removeItems(bKey)
        self.refresh3d()
   ```

### Erratic bug on `difference()`
Sometime, we have this exception

  ```
  File "C:\platform\python\Miniconda3\envs\3d\lib\site-packages\madcad\triangulation.py", line 289, in triangulation_outline
    raise TriangulationError("no more feasible triangles (algorithm failure or bad input outline)", [outline.indices[i] for i in hole])
madcad.triangulation.TriangulationError: ('no more feasible triangles (algorithm failure or bad input outline)', [4, 21, 19])
ERROR conda.cli.main_run:execute(33): Subprocess for 'conda run ['python', 'c:/jean-louis/projets/test/0-python/images/3d/test_madCAD.py']' command 
failed.  (See above for error) 
  ```

**Comment by jimy-byerley** [comment]
Hello @jlvila !

There seems to be 2 unrelated issues here:

## wrong side suppression issue
I reformated your original text to highlight the code, so it is more readable.
Looking at the problem, the boolean operation `difference` seems to take the wrong side from `brickA` (the side of `brickA` that is inside `tubeB`) this is often an issue with the outer normal of the other mesh, so your `tubeB` must have the outer normals inside instead of outside as you expect.  
This could be just a wrong sign with the direction of the axis of the tube, or with the extrusion direction.

Can you put your code for generating the cylinder and the bricks ?

## `TriangulationError` issue

Well as it says, the boolean operation fails to retriangulate the surfaces after intersection of one mesh with the other. Here it happens because the intersection doesn't give consistent intersection cuts, for a simple reason:   a portion of the small brick is both *in-* and *out-side* of the big one.

![ambiguity](https://user-images.githubusercontent.com/7940753/157922528-7f237c7a-ebfd-46bf-8cf3-6adb7806f41f.png)

For my current implementation of the boolean operations, this is an ambiguity that is not supported.
To raise the ambiguity and workaround the issue, you can just make the small brick slightly bigger (just a `1e-4` offset of the ambiguous surface is sufficient). Check #8 for details.

Hope this help !

**Comment by jlvila** [comment, marked as answer]
Hi Jimy,

### Difference
Thanks, for tube unexpected difference, I just change, as you suggested, the direction of the axis, and the result is as expected ;-)
I change `vec3(0, 1, 0)` to `vec3(0, -1, 0)` for the _circle.

#### The code
```python
    def addTube(self, x, y, w, h, color: QColor):
        _radius = w/2.0
        _lenght = h
        _circle = flatsurface(wire(Circle((vec3(0), vec3(0, -1, 0)), _radius)))
        _tube = Tube3D(_circle, [
            vec3(0, 0, 0),
            vec3(0, _lenght, 0),
        ])
        brTranslated = _tube.transform(
            vec3(x+_radius, y, 0))
        brTranslated.option(color=vec3(color.red()/255.0,
                                       color.green()/255.0, color.blue()/255.0))
        _key = self.scene3d.scene.add(brTranslated)
        npart4 = note_bounds(brTranslated)
        self.scene3d.scene.add(npart4, f'{_key}_bounds')
        print(f'Add tube from 3D: {_key}')
        self.refresh3d()
        return brTranslated, _key
```

#### `TriangulationError `issue
I just add a randomized float value to Z to drastically minimize the risk of this `TriangulationError`.
